trie/hexary.py

```diff
@@ -46,18 +50,18 @@
 assert BLANK_HASH == keccak(b'')
 
 
-class HexaryTrie(object):
-    db = None
-    root_hash = None
+class HexaryTrie:
+    __slots__ = ('db', 'root_hash', 'is_pruning')
 
     # Shortcuts
     BLANK_NODE_HASH = BLANK_NODE_HASH
     BLANK_NODE = BLANK_NODE
 
-    def __init__(self, db, root_hash=BLANK_NODE_HASH):
+    def __init__(self, db, root_hash=BLANK_NODE_HASH, prune=False):
```

Comment:
Just curious if you have any idea what the performance impact of using `prune=True` would be. Do you think it'd make sense for us to use that by default in py-evm or should we just use the `squash_changes()` contextmanager when generating tx/receipt tries?

Reply:
Seems like pruning on by default would be a good thing.

Reply:
Fair, and I think we do actually make this assumption of being able to roll back to previous state roots in py-evm.  

I'm still inclined to make this the default behavior as the current behavior is not really documented and our use of it is largely *implicit* based on the implementation.  Maybe it's worth doing some minor experimentation after releasing this with it off by default to see how bad it changes things by turning it on by default within py-evm.